python/ray/_private/gcs_pubsub.py

```diff
@@ -139,63 +212,138 @@ def subscribe_error(self) -> None:
-                    try:
-                        fut.result(timeout=1)
-                        break
-                    except grpc.FutureTimeoutError:
-                        # Subscriber has closed. Cancel inflight the request
-                        # and return from polling.
-                        if self._close.is_set():
-                            fut.cancel()
-                            return None, None
-                        # GRPC has not replied, continue waiting.
-                        continue
-                    except Exception:
-                        # GRPC error, including deadline exceeded.
-                        raise
+            req = self._subscribe_logs_request()
+        self._stub.GcsSubscriberCommandBatch(req, timeout=30)
+
+    def _do_polling(self) -> grpc.Future:
+        req = self._poll_request()
+        fut = self._stub.GcsSubscriberPoll.future(req)
+        while True:
+            # Terminate the polling if the subscriber has closed.
+            if self._close.is_set():
+                break
+            try:
+                # Release lock and wait for at most 1s for result to
+                # become available. This allows checking if the subscriber
+                # has closed.
+                fut.result(timeout=1)
+                # Result available.
+                break
+            except grpc.FutureTimeoutError:
+                # GRPC has not replied, continue waiting.
+                continue
+            except Exception:
                 if fut.done():
-                    for msg in fut.result().pub_messages:
-                        self._errors.append((msg.key_id,
-                                             msg.error_info_message))
+                    # This should be a GRPC error e.g. connection unavailable.
+                    # Abort and retry from the outer loop.
+                    # The exception is stored in fut.
+                    break
+                # Unknown failure.
+                raise
+        return fut
+
+    def _polling_loop(self):
+        """Main loop of the polling thread.
+
+        Polling thread is terminated when subscriber has closed.
+        """
+        exception_count = 0
+        while True:
+            # Terminate the thread if the subscriber has closed.
+            if self._close.is_set():
+                break
+
+            # Run one polling request until it is done or an error occurs.
+            fut = self._do_polling()
+
+            # Happens only when the subscriber is closed.
+            if fut.running():
+                fut.cancel()
+                break
+
+            # Ignore failure and retry polling.
+            if fut.exception():
+                exception_count += 1
+                if exception_count % 100 == 0:
+                    logger.warn("GCS subscriber polling failed: "
+                                f"{fut.exception()}. Retrying ...")
+                time.sleep(1)
+                continue
```

Comment:
This seems pretty aggressive. Why not give up after a few retries (or not do retries)?

Reply:
The polling were done in a separate thread, and we wanted to have the polling thread running iff subscriber is not closed. Refactored to run polling by the poller.